### ISSUE TITLE
Reimplement ms1-ms2 spectra linking (method 3) in python

### ISSUE THREAD

**Issue by joewandy** [sdrogers/ms2ldaviz#33]
It should be possible to use pymzml to do this.

Also check in the original method 3, how many MS2 peaks are we actually throwing away?

**Comment by sdrogers** [comment]
- Simon can write an mzml parser
 - Joe to add Method 3 to it

**Comment by sdrogers** [comment]
PS: speaking with Karen and Ronan yesterday, they would like to use this to replace the r method 3 code in frank. So, we should make this pretty high priority...

**Comment by sdrogers** [comment]
Hi again.
I've done the first bit (see LoadMZML object in ms2lda_feature_extraction in the lda repository).
This:
 - parses the mzml file
 - keeps track of the most recent ms1 scan
 - for each ms2 scan, looks at its precursor_mass and looks *back* at the most recent MS1 to find the most intense peak within plus/minus the isolation window
 - This peak (if found) is used to define the document
The current code outputs the list of ms1 and ms1 objects and metadata as normal. I envisage the following method 3:
 - Once the above has finished (but still in the load_spectra method (or at least called from there))
 - Load a peak list file
 - Match the peaks in the peaklist file to the MS1 objects (picking, say most intense if we have multiple ms1 matches to a peak in the peaklist within the tolerances (tolerance parameters are already there))
 - keep track of the subset of ms1 objects that are matched to something
 - Change their mz and rt to the value in the peaklist? **[do we do this now? If we don't, ignore this step]**
 - Make new ms1 and ms2 lists just including the ms1 subset and their ms2 objects

Does that sound reasonable? The bit I'm not sure of is what we do a) when there are multiple fragmented peaks that match the peaklist peak (I recall a couple of options in the R) and b) if, once we've decided on a match, whether or not we change the mz and rt values of the ms1 object to be exactly those in the peaklist. No idea if we do this now or not, probably not.

**Comment by justinjjvanderhooft** [comment]
Hi Simon,

Part 1 sounds very reasonable and good to me.

As for method 3 - the way it currently works is that the peak list is used as template to match MS/MS spectra to.within a certain tolerance window of RT.  
So the final m/z and RT values are always coming from the peak list indeed (so that they match correctly with MS1 (full scan) data, so the mz and rt values of the ms1 object are exactly those in the peak list.

As you rightly put there, there are several options in R to handle the situation of multiple MS/MS  spectra within a tolerance window. Currently, I opted for the most intense one (with the most intense base peak) - other options in R are the one closest in RT or  an 'average' spectrum.

I would suggest to try the 'traditional method 3' and the 'Python-based method 3' and see in how far the results are similar/the same.

Hope things are clear,

Cheers Justin

**Comment by sdrogers** [comment]
Ok, so can we agree on the following process?

 - For each MS2 spectrum in the file:
   - Find the most intense MS1 in the most recent MS1 scan within the isolation window of the quoted MS2 precursor value
   - If it can't find anything (not sure how this could happen), ignore the MS2
   - If it does find something, assume that this MS1 peak is the one that has been fragmented, and make a document
   - output is a list of ms1 objects, a list of ms2 objects (with refs to their respective ms1s and some metadata
 - If a peaklist exists:
   - For each peak in peaklist:
     - Find all the *MS1* objects within a mass and rt tolerance
     - if there aren't any, move to next peak in peaklist
     - if there are several, choose one, based on one of the criteria justin mentions above (i.e. maybe the one with the most intense MS1 peak, maybe the one with the most intense fragment base peak)
     - Make a new MS1 object with the mz and rt from the peaklist 
     - For each of the ms2 objects that pointed to the chosen ms1 object, make a new ms2 object identical except it now refers to the new ms1 object and not the original
     - remove the chosen ms1 object so that it can't be matched again [not sure this happens in the R code]?


In other words it there are two stages: find all the ms2 and the most likely ms1 that was fragmented. Match, where possible the peaks in the peaklist to one of the ms1 objects (and therefore the ms2)

If that sounds good, either Joe or I can implement that quickly....

**Comment by justinjjvanderhooft** [comment]
Okay. Do I get it right that in the first part, we might get several ms1 objects belonging to the same metabolite, i.e., if it got fragmented several times.

In the second part, for the ms1 - ms1 matching, we can now apply both a RT and mz tolerance and that should normally result in one matched peak. However, several ms1 objects from the fragmentation file might match to the same ms1 object in the peak list - then we could keep the one with the most intense ms2 object paired to it. 

Then we could try it on the four beers and analyze and compare the number of MS1-MS2 pairs as compared to the R method 3 (and compare the selected ms1 objects)

**Comment by sdrogers** [comment]
Exactly...
